### PR TITLE
feat: vulns exit code 1, errors 2

### PR DESCRIPTION
**Pull request by lili2311**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
- Add error code 2 for all failure states apart from *vulns found*, give's user a chance to act on this and retry if needed
- Stop treating *vulns* as an error that is treated as a `bad-command`, stop further processing and display vuln results.

#### Where should the reviewer start?
Start in command/test

#### How should this be manually tested?
Run snyk locally and try a few failed commands vs normal test runs

#### Any background context you want to provide?

We send all the vuln found commands back and pollute our analytics, it is hard to differentiate between actual failures in this way

MultipleTestsExistCodes2

### CODE REVIEW

File: src/lib/errors/legacy-errors.js [hunk 1]
```diff
@@ -71,10 +69,6 @@ module.exports = function error(command) {
 module.exports.message = function(error) {
   let message = error; // defaults to a string (which is super unlikely)
   if (error instanceof Error) {
-    if (error.code === 'VULNS') {
```
Comment (lili2311) [inline src/lib/errors/legacy-errors.js]: this is not set anywhere

File: src/lib/errors/legacy-errors.js [hunk 2]
```diff
@@ -46,7 +45,6 @@ const codes = {
   404: errors.notfound,
   411: errors.endpoint, // try to post to a weird endpoint
   403: errors.endpoint,
-  401: errors.auth,
```
Comment (lili2311) [inline src/lib/errors/legacy-errors.js]: this error was refactored into a new custom error, removing from legacy errors

File: src/cli/commands/test.ts [hunk 3]
```diff
@@ -162,11 +163,17 @@ async function test(...args): Promise<string> {
...
-    error.userMessage = (vulnerableResults[0] || errorResults[0]).userMessage;
+    error.code = errorResults[0].code;
+    error.userMessage = errorResults[0].userMessage;
     throw error;
   }
 
+  if (foundVulnerabilities) {
+    response += chalk.bold.red(summaryMessage);
+    console.error(response);
+    process.exit(1);
```
Comment (darscan) [inline src/cli/commands/test.ts]: I don't think we can exit like this.. due to things like `sln` and the multi test thing (`snyk test folder1 folder2 folder3`)

Comment (kyegupov) [reply src/cli/commands/test.ts]: I think both are handled ok: `for (const path of args) {` is done by the line 104 and sln.updateArgs already was applied to args.

File: src/cli/index.ts [hunk 4]
```diff
@@ -38,12 +38,9 @@ async function runCommand(args) {
...
-  let command = 'bad-command';
+  const command = 'bad-command';
 
-  if (error.code === 'VULNS') {
-    // this isn't a bad command, so we won't record it as such
-    command = args.command;
-  } else if (!error.stack) { // log errors that are not error objects
+  if (!error.stack) { // log errors that are not error objects
     analytics.add('error', JSON.stringify(error));
     analytics.add('command', args.command);
```
Comment (kyegupov) [inline src/cli/index.ts]: is "bad-command" actually of any use as the analytics data?

File: src/cli/index.ts [hunk 5]
```diff
@@ -55,6 +52,10 @@ async function handleError(args, error) {
     if (error.message && error.message.vulnerabilities) {
       delete error.message.vulnerabilities;
     }
+    if (error.vulnerabilities) {
+      delete error.message.vulnerabilities;
```
Comment (kyegupov) [inline src/cli/index.ts]: delete error.vulnerabilities

File: src/cli/index.ts [hunk 6]
```diff
@@ -132,15 +133,15 @@ async function main() {
   }
 
   if (!process.env.TAP && failed) {
-    process.exit(1);
+    process.exit(2);
```
Comment (kyegupov) [inline src/cli/index.ts]: 1. Maybe extract EXIT_CODE_VULNS and EXIT_CODE_ERRORS?

2. These need to be documented in the main README.md of the repo

File: src/cli/index.ts (hunk 4 above)
Comment (lili2311) [reply src/cli/index.ts]: yes it is every actual error + vulns and we have dashboardsrelying on this

File: src/cli/commands/test.ts (hunk 3 above)
Comment (lili2311) [reply src/cli/commands/test.ts]: Will verify, also need to test `sln` which relies on multiple results back too

Comment (lili2311) [reply src/cli/commands/test.ts]: Multiple test runs works :D Will document it all and add screenshots / gifs. Need to check SLN

Comment (lili2311) [reply src/cli/commands/test.ts]: That also seems to work, will look into the tests

**Pull request closed without merging**